docs/DesignDocs/Enabledebugmalloc.md

```diff
@@ -0,0 +1,229 @@
+
+- Whenever memory is freed, it clears the contents to a fixed pattern to help detect access of freed memory.
+  It also frees the associated callstack information.
+
+- When enclave terminates, it prints information about leaks - those objects that have been allocated,
+  but not feed.
+
+Debug malloc is performance intensive and is therefore not compiled into the SDK release.
+
+There have been requests from users that debug malloc be made available so that they can
+identify leaks within the enclave. Commonly used memory detection tools like Valgrind do not
+work within the enclaves since they cannot access enclave memory.
+
+According to the user's developing experience, there are two major problems
+which the memory leaking detection can help to address:
+
+- A developer wants to track all allocations from the start of the enclave
+  until it is terminated and determine if any allocations were leaked.
+  We name it "global detection".
+
+- A developer wants to track allocations in a specific region of the enclave
+  app code that they can specify at dev time. We name it "local detection".
+
+So far the "global detection" has already been fulfilled and improved across
+several pull requests, then this doc will emphasize the design of "local
+detection".
+
+# User Experience
+
+## Global detection
+
+### Default option
+
+The default option is "USE_DEBUG_MALLOC". Now we already set "USE_DEBUG_MALLOC"
+to "ON" in debug mode by default. Users can modify this default option
+according to the need, such as change it to "ON" in release mode or change it
+to "OFF" in debug mode.
+
+### Customed option
+
+Debug malloc will be shipped to the users as a separate library
+oelibdebugmalloc.a, so users can turn on debug malloc per enclave while the
+default option is "OFF".
+
+To use debug malloc, users will have to link oelibdebugmalloc.a into their
+enclaves explictly. This can be done by
+```
+      cmake:
+      target_link_libraries(user-enclave oelibc oedebugmalloc oeenclave)
+
+      make:
+      LDFLAGS=$(shell pkg-config oeenclave-$(COMPILER) --libs)
+      LDFLAGS_CUSTOM=$(subst -loelibc, -loedebugmalloc -loelibc, $(LDFLAGS))
+      $(CC) -o user-enclave user-enc.o $(LDFLAGS_CUSTOM)
+
+```
+
+For the global detection, if there is any memory leak during the enclave's
+life circle, an error code "OE_MEMORY_LEAK" will be raised during the enclave
+termination. This means it not only can check the users' enclave application,
+but also can cover the SDK's allocation of the enclave side.
+
+## Local detection
+
+Currently, debug malloc prints leaks only during enclave termination.
+However, users would like to isolate and track specific memory allocations.
+To achieve this, the following APIs are provided in a new published header `debugmalloc.h`.
+
+- `bool oe_debug_malloc_tracking_start(void)`
+  Start tracking memory allocations.
+
+- `bool oe_debug_malloc_tracking_stop(void)`
+  Stop tracking memory allocations.
```

Comment:
What happens if you call oe_debug_malloc_tracking_stop() when USE_DEBUG_MALLOC is set?  Is that an error?

Reply:
oe_debug_malloc_tracking_stop() is available when USE_DEBUG_MALLOC  is "on" or oedebugmalloc.a is linked explicitly by hand. So there is no error due to compiling option.

Reply:
The "Global detection" section of this doc does not mention the stop API, and the Local detection section says "- oe_debug_malloc_tracking_stop() can only been called after calling
oe_debug_malloc_tracking_start()".   So if it can be called in Global detection mode, then the doc should be updated to say so.

Reply:
Thanks, it is updated to state global detection is independent of these APIs.